src/plugins/oauth.js

```diff
@@ -1,29 +1,19 @@
 import qs from 'querystring';
 
-const TOKEN_REFRESHED = 'TOKEN_REFRESHED';
-const TOKEN_REFRESH_FAILED = 'TOKEN_REFRESH_FAILED';
-
 export default {
   async onStart(request) {
     if (this.client.refreshing) {
-      request.secondTry = true;
-
-      await new Promise((resolve) => {
-        this.client.eventEmitter.once(TOKEN_REFRESHED, () => {
-          request.headers.set(
-            'Authorization',
-            `Bearer ${this.client.getBearerToken()}`
-          );
-          resolve();
-        });
-        this.client.eventEmitter.once(TOKEN_REFRESH_FAILED, () => resolve());
-      });
-    } else {
-      request.headers.set(
-        'Authorization',
-        `Bearer ${this.client.getBearerToken()}`
+      const refreshed = await new Promise(
+        (resolve) => (this.client.refreshed = resolve)
```

Comment:
I haven't run this to see, so correct me if I'm wrong:

If you have multiple requests coming in while an id token is being refreshed then `this.client.refreshed` will be overwritten for each one of those incoming requests. When the refresh is completed and `this.client.refreshed()` is called only the last request in the queue will be resolved, leaving the previous requests permanently unresolved.

IIRC we tried writing tests for cases like this, but it's difficult to test for

Reply:
Is the desired behavior that the subsequent requests wait until the refresh occurs, and then proceed? Instead of instantiating a new promise in each `onStart` we could check for any existing promise and await that.

Reply:
Yeah.

Basically the fetch-client is handling a stream of requests using the authorization token that it knows about. One of those requests will eventually return a 401 because the token has expired, and and the oauth plugin goes and gets a new token. While the oauth plugin is doing this work the stream of requests is still coming in, and they will all return a 401 if we don't pause them and wait for the refresh to complete.

We've been emulating a queue of requests using the event handler, essentially. During the refresh we don't execute requests right away, we add event handlers that execute the requests to the queue, and when the "refreshed" event fires all of those handlers are called. It's pretty convoluted frankly.

I think checking for any existing promise and awaiting that seems like it might work, but I'm worried that it'd be possible to introduce a race condition here. I think we tried something like that and it turned out to be hard. IIRC `this.client.refreshPromise` was a thing we set and tried awaiting